src/socket.rs

```diff
@@ -52,28 +52,21 @@ impl NlSocket {
 
     /// Send message encoded as byte slice to the netlink ID specified in the netlink header
     /// (`nl::nlhdr::NlHdr`).
-    pub fn send(&mut self, buf: &[u8], flags: i32) -> Result<isize, io::Error> {
+    pub fn send(&mut self, buf: &[u8], flags: i32) -> Result<usize, io::Error> {
         match unsafe {
             libc::send(self.fd, buf as *const _ as *const c_void, buf.len(), flags)
         } {
-            i if i >= 0 => Ok(i),
+            i if i >= 0 => Ok(i as usize),
             _ => Err(io::Error::last_os_error()),
         }
     }
 
     /// Receive message encoded as byte slice from the netlink socket.
-    pub fn recv(&mut self, len: Option<usize>, flags: i32) -> Result<Vec<u8>, io::Error> {
-        let mut v = match len {
-            Some(l) => vec![0; l],
-            None => Vec::new(),
```

Comment:
This part was likely a bug unless I'm missing the use-case? `Vec::new` will return an empty vector and thus `v.as_mut_slice()` will be `[]` and have length zero. Thus it can never read any byte into that slice if `None` is passed for `len`.

Reply:
Yes that seems to be a bug. If the user is required to allocate it from the application, that should remove the need for the size parameter as long as the proposed change is a fat pointer or something similar. One thing I will warn you about with netlink however is that if I recall correctly, netlink will drop any piece of the message that is not captured in a single read (the original reason I decided to use a `Vec`). Just keep that in mind when you make changes for the API as that's something that could definitely cause issues. This, however, might just be a documentation problem as opposed to something that needs to be guarded against in the code.

Reply:
This does not change how much a user can read. The length of the buffer is still fully controllable by the user.

Reply:
Just to clarify, by issues I mean for the user, not from a technical standpoint. I'm happy to use a fat pointer but I am more weighing options in my head here with regard to where the safeguard should be here: in the docs warning about data loss with improperly sized buffers or in the code and the more I think about it I tend to agree that more control here is probably better.

Reply:
Yes. Makes sense to say in the documentation that netlink will discard any part of a message that does not fit in the buffer. Linux does the same thing for UDP as well (completely unrelated).